### PR TITLE
Extended Membership Protocol

### PR DESCRIPTION
**Pull request by SebastianStehle**

### CODE REVIEW

**Comment by SebastianStehle** [comment]
The membership provider supports three different strategies now:

1. ```Single```: A single document per deployment. Fastest for small clusters.
2. ```Multiple```: One document per silo and an extra document for the table version. Needs a replica set and transaction support to work properly.
3. ```MultipleDeprecated```: One document per silo but no support for the extended membership protocol. 

I have also copied the newest binaries to the repo to use the TesterInternal tools.

File: Orleans.Providers.MongoDB/Membership/Store/Multiple/MultipleMembershipCollection.cs [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+                    var tableVersion = tableVersionCollection.GetTableVersionAsync(deploymentId);
+
+                    var entries =
+                        Collection.Find(x => x.DeploymentId == deploymentId)
+                            .ToListAsync();
+
+                    await Task.WhenAll(tableVersion, entries);
+
+                    return ReturnMembershipTableData(entries.Result, tableVersion.Result);
```
Comment (ReubenBond) [inline Orleans.Providers.MongoDB/Membership/Store/Multiple/MultipleMembershipCollection.cs]: Should there be a `CommitTransactionAsync` call here and elsewhere?

Comment (SebastianStehle) [reply Orleans.Providers.MongoDB/Membership/Store/Multiple/MultipleMembershipCollection.cs]: I don't think so, as we do not make any updates.

Comment (ReubenBond) [reply Orleans.Providers.MongoDB/Membership/Store/Multiple/MultipleMembershipCollection.cs]: That seems odd - is the transaction just hanging in that case or does it automatically abort when the session scope ends?

Comment (SebastianStehle) [reply Orleans.Providers.MongoDB/Membership/Store/Multiple/MultipleMembershipCollection.cs]: You are probably right, the end session will stop it, but it is blocking. So I think it is better to commit manually.

**Pull request merged**